### ISSUE TITLE
Caching for php/ composer projects?

### ISSUE THREAD

**Issue by andig** [travis-ci/travis-ci#4579]
With the new container infrastructure, caching has become available for public repos (http://docs.travis-ci.com/user/migrating-from-legacy/). The caching docs (http://docs.travis-ci.com/user/caching/) currently don't mention php specifically.

I'm using this setup but am unsure if this actually works:

    cache:
      directories:
        - $COMPOSER_CACHE_DIR
        - $HOME/.composer/cache
        - vendor

These should be (?) pretty much the default directories for php using composer. If yes- should it be added to the docs for benefit of other users?

**Comment by stof** [comment]
Caching the ``vendor`` itself looks like a bad idea to me, because it can create nasty side effects.

Caching ``$HOME/.composer/cache`` should be enough (or even only ``$HOME/.composer/cache/files`` to persist the cached downloaded archives but not the packagist metadata as they are likely to change anyway on each build and invalidate the cache too often)

**Comment by andig** [comment]
> Caching the vendor itself looks like a bad idea to me, because it can create nasty side effects.
> Caching $HOME/.composer/cache should be enough

Yep, saw another post on the same note. It should be added as best practice to the docs I'd suggest.

**Comment by BanzaiMan** [comment]
Is document update sufficient? If so, could you look at https://github.com/travis-ci/docs-travis-ci-com and suggest changes? Thank you!

**Comment by stof** [comment]
@BanzaiMan given there is nothing related to composer in the builtin steps (as https://github.com/travis-ci/travis-build/pull/278 has not been merged yet), this is indeed a documentation-only change

**Comment by andig** [comment]
If there are no objections I could update the docs to suggest:

  - composer: $HOME/.composer/cache
  - npm: $HOME/.npm

Does that sound sensible?

**Comment by stof** [comment]
for composer, I would use only ``$HOME/.composer/cache/files``.
The ``repo`` subfolder in the cache contains metadata downloaded from Packagist (and other repos you may have registered in the project), and these would probably have changed for each build (the main files giving access to others are updated every few minutes because of the frequency of changes in packages). I'm not sure the time won thanks to the caching of some of the metadata files (the ones for packages not updated since your previous Travis build) would actually outweight the time spent storing the Travis cache again for each build.
The ``file`` subfolder on the other hand stores the archives downloaded with the source code, and you have a real benefit here (and there is a good chance to hit the cache, especially if you depend on stable versions of packages).

**Comment by acoulton** [comment]
It looks like if you only cache `$HOME/.composer/cache/files` it only caches downloaded archives, not vendors that have been cloned. This means it's not actually caching much, because it's very common that travis has been rate-limited by github and has to fall back to cloning when downloading an archive fails.

I **think** there's no way to cache the cloned dependencies without adding `/vendor` to the cache list?

**Comment by andig** [comment]
>  because it's very common that travis has been rate-limited by github and has to fall back to cloning when downloading an archive fails.

Does that really make sense? If we're caching the archives then downloading those should no longer be failing due to rate limits?

**Comment by andig** [comment]
>  because it's very common that travis has been rate-limited by github and has to fall back to cloning when downloading an archive fails.

If we're caching the archives then downloading those should no longer be failing due to rate limits?

**Comment by acoulton** [comment]
That assumes at least one build is able to download the archives to populate the cache. So far on the project that I've tweaked the cache directories on, it's failed to download every archive on every run so the cache is empty.

Yes, eventually you may get lucky and get a few archives into your cache and then of course you'll be fine on subsequent builds of that branch until the cache is cleared/refreshed.

**Comment by andrerom** [comment]
cache is enabled now on all versions of travis, and even for public builds it seems: https://blog.travis-ci.com/2013-12-05-speed-up-your-builds-cache-your-dependencies

As pointed out by @stof we should not cache vendor folder. The good advice would be to cache `$HOME/.composer/cache/files` and setup a personal/org github auth token *(a read only one with no rights what so ever, and ideally connected to a user without rights as well to be on the safes side..)*, that will cache archives and avoid that you hit the rate limit. the right way TM ;)

**Comment by stof** [comment]
@andrerom not even needed. Downloads are not counted in the API rate limit anymore by github. They changed this a few months ago. The composer team contacted their support to find the best way to workaround the rate limiting issues while still using supported endpoints, and their infrastructure team said that their new download system was performant enough to exclude it from the very low anonymous rate limit of the API.

**Comment by BanzaiMan** [comment]
So, the consensus is to cache `~/.composer/cache/files` for Composer?

**Comment by andrerom** [comment]
Seems so, but maybe @Seldaek or @naderman can give a official recommendation.

**Comment by andrerom** [comment]
On the other hand composer itself is caching the whole cache folder: https://github.com/composer/composer/blob/master/.travis.yml

**Comment by naderman** [comment]
@andrerom thanks for pointing that out, should probably change that as it looks like it updates the cache nearly every build atm.

It's indeed the case that the cache contains a few files in `.composer/cache/repo/` that change very often as @stof said. There is a folder that can be cached for cloned repositories, which is `.composer/cache/vcs`. So considering the performance impact of generating an entire new archive and uploading it since there is no way to update just those files, I would recommend the cache be set up for `.composer/cache/files` and `.composer/cache/vcs` (@stof any reason you didn't include the vcs directory?)

**Comment by stof** [comment]
@naderman most projects I'm working on are not cloning repositories but only downloading archives, and so would not use it.
Thus, 2 years ago when I settled on this, an install from source would not use the vcs cache. Only registering a VCS repository in the project would, so this made it even less worthy at that time.
But it may indeed make sense to include it (but we may face the same issue about regularly changing files though, so not sure)

**Comment by xiaohuilam** [comment]
And anyone knows how to cache `composer.lock`, to accelerate composer install?
